packages/adapter-utils/src/compare.ts

```diff
@@ -396,21 +390,22 @@ export const applyDetailedChanges = (element: ChangeDataType, detailedChanges: D
   )
 
   const [realListItemGroup, otherGroups] = _.partition(Object.values(potentialListItemGroups), group =>
-    Array.isArray(resolvePath(element, group[0].id.createParentID())),
+    Array.isArray(resolvePath(element, getChangeRealId(group[0]).createParentID())),
   )
 
   _(otherChanges)
     .concat(otherGroups.flat())
+    .filter(filterFunc)
     .forEach(detailedChange => {
-      const id = isRemovalChange(detailedChange)
-        ? detailedChange.elemIDs?.before ?? detailedChange.id
-        : detailedChange.elemIDs?.after ?? detailedChange.id
+      const id = getChangeRealId(detailedChange)
       const data = isRemovalChange(detailedChange) ? undefined : detailedChange.data.after
       setPath(element, id.replaceParentId(element.elemID), data)
     })
 
-  realListItemGroup.forEach(changes => {
-    applyListChanges(element, changes)
+  // we want inner lists to be applied before outer lists, because the indexes may change
+  const orderedListItemGroups = _.orderBy(realListItemGroup, group => group[0].id.nestingLevel, 'desc')
+  orderedListItemGroups.forEach(changes => {
+    applyListChanges(element, changes, filterFunc)
```

Comment:
```suggestion
  _.orderBy(realListItemGroup, group => group[0].id.nestingLevel, 'desc').forEach(changes => {
    applyListChanges(element, changes, filterFunc)
```
Just a matter of preference, take it or leave it.